feat: add test data, fix peaksData and add show method

jorainer opened the pull request:
- Add test data file.
- Fix `peaksData` for duplicated spectra.
- Add `show` method.

jorainer committed:
feat: add test data, fix peaksData and add show method

- Add test data file.
- Fix `peaksData` for duplicated spectra.
- Add `show` method.

jorainer commented on R/MsBackendTimsTof-functions.R:
```diff
@@ -138,16 +144,50 @@ MsBackendTimsTof <- function() {
-            res[I] <- unname(split(tmp[, sd], f))
-        else res[I] <- unname(split.data.frame(as.matrix(tmp[, sd]), f))
+        ## Ensure mz and intensity are always first, if provided
+        sd <- c(intersect(c("mz", "intensity"), sd),
```
This is to ensure that `"mz"` and `"intensity"` are always (if requested) reported as the first two columns.

andreavicini replied:
Just out of curiosity, is there a specific reason for having them as first columns?

jorainer replied:
It's for compatibility reasons: there is the unwritten rule that the first column should contain m/z and the second intensity values. While in `Spectra` we access the columns with their actual name, it could be that other developers rely on that. So, this is just to ensure that.

jorainer replied:
Maybe I should however mention that in the documentation?

andreavicini replied:
Mentioning it in the small documentation above the function wouldn't be wrong I guess but for me it's ok also like it is, it was just a curiosity :) . Should also `spectraData` preserve this order then?

jorainer replied:
no, for `spectraData` it does not really matter. That should be more like a `data.frame` and then the columns should be returned in the order they are requested. I will add a short documentation to that.

jorainer commented on R/MsBackendTimsTof-functions.R:
```diff
@@ -138,16 +144,50 @@ MsBackendTimsTof <- function() {
+        i_frame <- x@indices[I, "frame"]
+        tmp <- query(tms, unique(i_frame), c("frame", "scan", sd))
+        ## subset tmp if we're about to extract only few scans.
+        if (length(i_frame) < (nrow(tmp) / 10)) {
```
This `if` statement subsets `tmp` if only few scans are to be extracted. This can result in considerably faster `split.data.frame` and `paste` calls below.

andreavicini replied:
I'm not sure I understand this. Basically I would say that the condition means that if the average number of scans per frame is greater than 10 we subset `tmp` to only contain scans that are in the backend. Is that wrong?

jorainer replied:
the `query` function allows only to retrieve data filtering by frame but not by spectrum. Thus, even if we want to extract a single spectrum we need to query all spectra in a frame and then we have that potentially large `data.frame` we need to split. I realized that it can be much faster if in such cases we first subset the `data.frame` to scan IDs we're interested in and then split it.

So, if we extract at least 10 times less scans then we have data, we first subset. Could also be that this is an overoptimization - or even wrongly implemented?

andreavicini replied:
I'm sorry I got confused erroneously thinking `length(i_frame)` as the number of frames but it is actually the number of scans. So basically we are subsetting when it is advantageous. I think it makes sense! Thank you!

jorainer replied:
yes, I know - it's confusing :)

jorainer commented on R/MsBackendTimsTof-functions.R:
```diff
@@ -138,16 +144,50 @@ MsBackendTimsTof <- function() {
+            rownames(tmp) <- NULL
+            ids <- paste(i_frame, i_scan)
+        } else ids <- paste(i_frame, x@indices[I, "scan"])
+        if (!nrow(tmp))
```
This is to avoid a problem if no peaks was recorded for **any** of the requested scans (spectra): without that a `list` with `logical` matrices is returned (instead of `numeric`).

jorainer commented on R/MsBackendTimsTof-functions.R:
```diff
@@ -138,16 +144,50 @@ MsBackendTimsTof <- function() {
+        if (!nrow(tmp))
+            tmp <- rbindFill(tmp, data.frame(frame = 0L))
+        f <- factor(paste(tmp$frame, tmp$scan), levels = unique(ids))
+        if (anyDuplicated(ids)) {
```
This is to ensure the function also supports `be[c(1, 2, 1, 2)]`, i.e. duplicated spectra.

andreavicini reviewed (commented):
Thanks @jorainer for the updates! Everything seems good to me! There's only one point that is not completely clear to me and I have put a quastion there.

jorainer committed:
docs: expand documentation on peaksData

- Clearly explain in the `peaksData` help that columns `"mz"` and `"intensity"`
  are **always** the first two columns to be returned (if requested).

jorainer commented:
I've now expanded the documentation on `peaksData`.

andreavicini reviewed (approved):
Perfect! Thanks @jorainer!
